File: src/stockscout_eod/contracts.py
"""Versioned wire contracts shared by the scanner, Pages PWA, and MCP index.

The public contract is intentionally smaller than the internal ``Candidate``
model. Full derived candidate evidence remains available in the detail shard;
the bounded chart contract is separate from candidate assets, while provider
caches and operator state remain forbidden.
"""
from __future__ import annotations

from typing import Any, Literal

from pydantic import AliasChoices, BaseModel, ConfigDict, Field

SCHEMA_VERSION = "stockscout-eod/v1"
MANIFEST_VERSION = 1

TradeStatus = Literal[
    "entry_ready",
    "trigger_pending",
    "wait_for_retest",
    "not_tradeable",
    "insufficient_data",
]


class WireModel(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class TradePlanV1(WireModel):
    status: TradeStatus
    reason_codes: list[str] = Field(default_factory=list, alias="reasonCodes")
    trigger_state: Literal["pending", "fresh", "extended", "unavailable"] = Field(
        alias="triggerState"
    )
    trigger_reference_level: float | None = Field(None, alias="triggerReferenceLevel")
    entry_reference_level: float | None = Field(None, alias="entryReferenceLevel")
    structural_invalidation_level: float | None = Field(
        None, alias="structuralInvalidationLevel"
    )
    entry_risk_pct: float | None = Field(None, alias="entryRiskPct")
    extension_atr: float | None = Field(None, alias="extensionAtr")
    tactical_stop_level: float | None = Field(None, alias="tacticalStopLevel")
    tactical_risk_pct: float | None = Field(None, alias="tacticalRiskPct")
    source: str
    version: int


class CandidateSummaryV1(WireModel):
    id: str
    mode: Literal["bottom-fishing", "next", "ryan-original"] | None = None
    price_basis: Literal["split_only", "split_div"] | None = Field(None, alias="priceBasis")
    canonical_url: str = Field(alias="canonicalUrl")
    ticker: str
    as_of: str = Field(alias="asOf")
    scan_order: int = Field(alias="scanOrder", ge=0)
    score: float
    focus_score: float | None = Field(None, alias="focusScore")
    focus_blend: float = Field(alias="focusBlend")
    headline_rank: int | None = Field(None, alias="headlineRank")
    price: float | None = None
    sector: str | None = None
    industry: str | None = None
    actionability: str
    primary_setup: str | None = Field(None, alias="primarySetup")
    setup_names: list[str] = Field(default_factory=list, alias="setupNames")
    setup_tags: list[str] = Field(default_factory=list, alias="setupTags")
    setup_states: dict[str, str | None] = Field(default_factory=dict, alias="setupStates")
    trade_plan: TradePlanV1 | None = Field(None, alias="tradePlan")
    trade_status: TradeStatus | None = Field(None, alias="tradeStatus")
    entry_risk_pct: float | None = Field(None, alias="entryRiskPct")
    tactical_stop_level: float | None = Field(None, alias="tacticalStopLevel")
    risk_level: str = Field("none", alias="riskLevel")
    risk_reasons: list[dict[str, str]] = Field(default_factory=list, alias="riskReasons")
    excluded: bool = False
    excluded_reason: str | None = Field(None, alias="excludedReason")
    data_status: str = Field("OK", alias="dataStatus")
    legacy_confirmation_status: str = Field("UNAVAILABLE", alias="legacyConfirmationStatus")
    setup: str | None = None
    setup_match_count: int = Field(0, alias="setupMatchCount")
    opportunity_score: float = Field(alias="opportunityScore")
    opportunity_rank: int = Field(alias="opportunityRank")
    opportunity_tier: str = Field(alias="opportunityTier")
    stage: int | None = None
    stage_name: str | None = Field(None, alias="stageName")
    rs_rank: float | None = Field(None, alias="rsRank")
    rs_score: float | None = Field(None, alias="rsScore")
    rs_3m: float | None = Field(None, alias="rs3m")
    rs_6m: float | None = Field(None, alias="rs6m")
    rs_new_high: bool | None = Field(None, alias="rsNewHigh")
    change_20d: float | None = Field(None, alias="change20d")
    return_3m: float | None = Field(None, alias="return3m")
    return_6m: float | None = Field(None, alias="return6m")
    volume_ratio: float | None = Field(None, alias="volumeRatio")
    avg_dollar_volume_50: float | None = Field(None, alias="avgDollarVolume50")
    atr_pct: float | None = Field(None, alias="atrPct")
    distance_50: float | None = Field(None, alias="distance50")
    distance_200: float | None = Field(None, alias="distance200")
    distance_30w: float | None = Field(None, alias="distance30w")
    from_52w_high: float | None = Field(None, alias="from52wHigh")
    trend_template_passes: int | None = Field(None, alias="trendTemplatePasses")
    base_weeks: int | None = Field(None, alias="baseWeeks")
    base_depth_pct: float | None = Field(None, alias="baseDepthPct")
    breakout_pct: float | None = Field(None, alias="breakoutPct")
    extended: bool = False
    sma50: float | None = None
    sma150: float | None = None
    sma200: float | None = None
    structure_score: float | None = Field(None, alias="structureScore")
    base_score: float | None = Field(None, alias="baseScore")
    trigger_score: float | None = Field(None, alias="triggerScore")
    accumulation_score: float | None = Field(None, alias="accumulationScore")
    long_base_score: float | None = Field(None, alias="longBaseScore")
    crash_base_score: float | None = Field(None, alias="crashBaseScore")
    rwb_squeeze_score: float | None = Field(None, alias="rwbSqueezeScore")
    ema_stack_launch_score: float | None = Field(None, alias="emaStackLaunchScore")
    ma_cluster_score: float | None = Field(None, alias="maClusterScore")
    sma_compression_pct: float | None = Field(None, alias="smaCompressionPct")
    ma_cluster_width_pct: float | None = Field(None, alias="maClusterWidthPct")
    fundamental_evidence_score: float | None = Field(None, alias="fundamentalEvidenceScore")
    fundamental_evidence_confidence: float | None = Field(
        None, alias="fundamentalEvidenceConfidence"
    )
    changed_today: bool | None = Field(None, alias="changedToday")
    new_universe_member: bool | None = Field(None, alias="newUniverseMember")
    change_impact: float | None = Field(None, alias="changeImpact")
    change_labels: list[str] = Field(default_factory=list, alias="changeLabels")


class CandidateDetailV1(CandidateSummaryV1):
    model_config = ConfigDict(extra="allow", populate_by_name=True)
    legacy_confirmation: dict[str, Any] = Field(
        default_factory=dict, alias="legacyConfirmation"
    )


class AssetDescriptorV1(WireModel):
    path: str
    sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    bytes: int = Field(ge=0)
    count: int = Field(ge=0)
    content_type: str = Field("application/json", alias="contentType")
    pattern: str | None = None
    bucket_count: int | None = Field(None, alias="bucketCount", ge=1)
    coverage_pct: float | None = Field(None, alias="coveragePct", ge=0, le=100)


class ScanCountsV1(WireModel):
    universe: int = Field(ge=0)
    candidates: int = Field(ge=0)
    excluded: int = Field(ge=0)
    failed: int = Field(ge=0)


class HealthCheckV1(WireModel):
    code: str
    passed: bool
    detail: str


class HealthV1(WireModel):
    status: Literal["healthy", "degraded", "failed"]
    coverage_pct: float = Field(alias="coveragePct", ge=0, le=100)
    checks: list[HealthCheckV1]


class ScanManifestV1(WireModel):
    manifest_version: Literal[1] = Field(MANIFEST_VERSION, alias="manifestVersion")
    schema_version: Literal["stockscout-eod/v1"] = Field(SCHEMA_VERSION, alias="schemaVersion")
    mode: Literal["bottom-fishing", "next", "ryan-original"] | None = None
    run_id: str = Field(alias="runId")
    session_date: str = Field(alias="sessionDate")
    market_data_date: str = Field(alias="marketDataDate")
    generated_at: str = Field(alias="generatedAt")
    status: Literal["healthy", "degraded"]
    price_mode: Literal["split_only", "split_div"] = Field(alias="priceMode")
    chart_status: Literal["ready", "stale", "missing"] = Field(
        "missing",
        alias="chartStatus",
        validation_alias=AliasChoices("chartStatus", "ownerChartStatus"),
    )
    counts: ScanCountsV1
    health: HealthV1
    provenance: dict[str, Any]
    versions: dict[str, str | int]
    assets: dict[str, AssetDescriptorV1]


class CloudFieldCatalogEntryV1(WireModel):
    field: str = Field(pattern=r"^[A-Za-z0-9_-]+(?:\.[A-Za-z0-9_-]+)*$")
    types: list[Literal["boolean", "null", "number", "string"]]
    count: int = Field(ge=0)
    example: Any = None


class CloudPublishCountsV1(WireModel):
    candidates: int = Field(ge=0)
    excluded: int = Field(ge=0)
    total: int = Field(ge=1)


class CloudPublishManifestV1(WireModel):
    """Control manifest accepted by the OIDC publish Edge Function.

    This is intentionally separate from :class:`ScanManifestV1`: the Pages
    pointer describes immutable assets, while this wrapper commits the exact
    candidate record set that Supabase will activate atomically.
    """

    schema_version: Literal["stockscout-eod/v1"] = Field(alias="schemaVersion")
    run_id: str = Field(alias="runId", pattern=r"^[A-Za-z0-9._:-]{1,100}$")
    scan_date: str = Field(alias="scanDate", pattern=r"^\d{4}-\d{2}-\d{2}$")
    market_data_date: str = Field(
        alias="marketDataDate", pattern=r"^\d{4}-\d{2}-\d{2}$"
    )
    generated_at: str = Field(alias="generatedAt")
    price_mode: Literal["split_only", "split_div"] = Field(alias="priceMode")
    health: HealthV1
    counts: CloudPublishCountsV1
    records_hash: str = Field(alias="recordsHash", pattern=r"^[0-9a-f]{64}$")
    manifest_hash: str = Field(alias="manifestHash", pattern=r"^[0-9a-f]{64}$")
    ranking_version: str | None = Field(None, alias="rankingVersion")
    detector_version: str | None = Field(None, alias="detectorVersion")
    provenance: dict[str, Any]
    asset_hashes: dict[str, str] = Field(alias="assetHashes")
    field_catalog: list[CloudFieldCatalogEntryV1] = Field(alias="fieldCatalog")


class ChartPayloadV1(WireModel):
    schema_version: Literal["stockscout-eod/chart-v1"] = Field(
        "stockscout-eod/chart-v1", alias="schemaVersion"
    )
    ticker: str
    as_of: str = Field(alias="asOf")
    price_mode: Literal["split_only", "split_div"] = Field(alias="priceMode")
    columns: list[str] = Field(default_factory=lambda: ["t", "o", "h", "l", "c", "v"])
    daily: list[list[int | float]] = Field(default_factory=list)
    weekly: list[list[int | float]] = Field(default_factory=list)


class ChartShardV1(WireModel):
    name: str
    sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    bytes: int = Field(ge=0)
    ticker_count: int = Field(alias="tickerCount", ge=0)


class ChartManifestV1(WireModel):
    schema_version: Literal["stockscout-eod/charts-v1"] = Field(
        "stockscout-eod/charts-v1", alias="schemaVersion"
    )
    run_id: str = Field(alias="runId")
    session_date: str = Field(alias="sessionDate")
    generated_at: str = Field(alias="generatedAt")
    price_mode: Literal["split_only", "split_div"] = Field(alias="priceMode")
    requested: int = Field(ge=0)
    available: int = Field(ge=0)
    coverage_pct: float = Field(alias="coveragePct", ge=0, le=100)
    storage_base_url: str = Field(alias="storageBaseUrl")
    shards: list[ChartShardV1]
    shards_by_ticker: dict[str, str] = Field(default_factory=dict, alias="shardsByTicker")


class RawScanEnvelopeV1(WireModel):
    schema_version: Literal["stockscout-engine/v1"] = Field(
        "stockscout-engine/v1", alias="schemaVersion"
    )
    run_id: str = Field(alias="runId")
    session_date: str = Field(alias="sessionDate")
    generated_at: str = Field(alias="generatedAt")
    price_mode: Literal["split_only", "split_div"] = Field(alias="priceMode")
    candidates: list[dict[str, Any]]
    excluded: list[dict[str, Any]] = Field(default_factory=list)
    stats: dict[str, Any]
    stage_rows: list[dict[str, Any]] = Field(default_factory=list, alias="stageRows")
    market: dict[str, Any] = Field(default_factory=dict)
    provenance: dict[str, Any] = Field(default_factory=dict)
    versions: dict[str, str | int] = Field(default_factory=dict)


def wire_dump(model: BaseModel) -> dict[str, Any]:
    return model.model_dump(mode="json", by_alias=True, exclude_none=True)

File: frontend/src/data/contracts.ts
export const STOCKSCOUT_SCHEMA='stockscout-eod/v1' as const
export const UNIFIED_SCHEMA='stockscout-unified/v1' as const
export type UnifiedManifestV1={
  manifestVersion:1;schemaVersion:typeof UNIFIED_SCHEMA;runId:string;sessionDate:string;generatedAt:string;status:'healthy';defaultMode:'bottom-fishing'
  modes:Record<'bottom-fishing'|'next'|'ryan-original',{mode:string;label:string;priceBasis:string;status:'healthy';manifestPath:string;manifestSha256:string;manifestBytes:number;candidates:number;excluded:number;chartCoveragePct:number;ranking:string}>
}

export type TradeStatus='entry_ready'|'trigger_pending'|'wait_for_retest'|'not_tradeable'|'insufficient_data'
export type TriggerState='pending'|'fresh'|'extended'|'unavailable'

export type TradePlanV1={
  status:TradeStatus
  reasonCodes:string[]
  triggerState:TriggerState
  triggerReferenceLevel:number|null
  entryReferenceLevel:number|null
  structuralInvalidationLevel:number|null
  entryRiskPct:number|null
  extensionAtr:number|null
  tacticalStopLevel:number|null
  tacticalRiskPct:number|null
  source:string
  version:number|string
}

export type AssetDescriptor={
  path:string
  sha256:string
  bytes:number
  count?:number
  coverage?:number
  coveragePct?:number
  pattern?:string
  shardCount?:number
  bucketCount?:number
}

export type ScanHealth='healthy'|'degraded'|'failed'

export type ScanManifestV1={
  manifestVersion:1
  schemaVersion:typeof STOCKSCOUT_SCHEMA
  mode?:'bottom-fishing'|'next'|'ryan-original'
  runId:string
  sessionDate:string
  marketDataDate:string
  generatedAt:string
  status:ScanHealth
  priceMode:'split_only'|'split_div'|string
  chartStatus?:'ready'|'stale'|'missing'
  ownerChartStatus?:'ready'|'stale'|'missing'
  counts:{universe:number;candidates:number;excluded:number;failed:number;total:number}
  health?:{status:ScanHealth;coveragePct:number;checks:Array<{code:string;passed:boolean;detail:string}>}
  provenance:Record<string,unknown>
  versions:{ranking:string;detectors:string;tradePlan:string;[key:string]:string}
  assets:{
    core:AssetDescriptor
    details:AssetDescriptor
    excluded:AssetDescriptor
    history:AssetDescriptor
    legacyConfirmation?:AssetDescriptor
    legacyIndex?:AssetDescriptor
    legacyDetails?:AssetDescriptor
    charts?:AssetDescriptor
  }
}

export type LegacyManifestV2={
  manifestVersion:2
  schemaVersion?:string
  model:string
  generatedAt:string
  runId?:string
  sessionDate?:string
  marketSession?:{date?:string|null;status?:string;timezone?:string}
  universe:number
  provenance:Record<string,unknown>
  assets:{
    core:AssetDescriptor
    legacyIndex:AssetDescriptor
    legacyDetails:AssetDescriptor
    legacyConfirmation:AssetDescriptor
    charts:AssetDescriptor
    details?:AssetDescriptor
    excluded?:AssetDescriptor
    history?:AssetDescriptor
  }
}

export type StockScoutManifest=ScanManifestV1|LegacyManifestV2

export type CandidateSummaryV1={
  id:string
  canonicalUrl?:string
  ticker:string
  mode?:'bottom-fishing'|'next'|'ryan-original'
  priceBasis?:'split_only'|'split_div'
  scanOrder:number
  focusBlend?:number
  tradeStatus?:TradeStatus
  entryRiskPct?:number|null
  tacticalStopLevel?:number|null
  excluded?:boolean
  [key:string]:any
}

export type CandidateDetailV1=CandidateSummaryV1&{
  tradePlan?:TradePlanV1|Record<string,unknown>|null
  setupHits?:Record<string,unknown>
  rawFeatures?:Record<string,unknown>
  reasons?:string[]
}

export type CandidateCoreV1={
  schemaVersion?:string
  runId?:string
  sessionDate?:string
  generatedAt:string
  market:Record<string,any>
  universe:CandidateSummaryV1[]
  detailShards?:Record<string,string>
  chartShards?:Record<string,string>
  [key:string]:any
}

export type ScanHistoryItemV1={
  runId:string
  sessionDate:string
  generatedAt:string
  status:ScanHealth
  coveragePct?:number
  candidateCount?:number
  excludedCount?:number
  counts?:ScanManifestV1['counts']
  manifestSha256?:string
}

function isRecord(value:unknown):value is Record<string,unknown>{
  return Boolean(value)&&typeof value==='object'&&!Array.isArray(value)
}

export function parseUnifiedManifest(value:unknown):UnifiedManifestV1{
  if(!isRecord(value)||value.schemaVersion!==UNIFIED_SCHEMA||value.manifestVersion!==1||value.status!=='healthy')throw new Error('Unified activation manifest is invalid')
  const modes=isRecord(value.modes)?value.modes:{}
  for(const mode of ['bottom-fishing','next','ryan-original'] as const){
    const pointer=modes[mode]
    if(!isRecord(pointer)||pointer.mode!==mode||pointer.status!=='healthy'||typeof pointer.manifestSha256!=='string'||!/^[0-9a-f]{64}$/.test(pointer.manifestSha256))throw new Error(`Unified mode pointer is invalid: ${mode}`)
  }
  return value as unknown as UnifiedManifestV1
}

function requiredText(value:unknown,label:string){
  if(typeof value!=='string'||!value.trim())throw new Error(`${label} is missing`)
  return value
}

function requiredAsset(value:unknown,label:string):AssetDescriptor{
  if(!isRecord(value))throw new Error(`${label} asset is missing`)
  return{
    ...value,
    path:requiredText(value.path,`${label}.path`),
    sha256:requiredText(value.sha256,`${label}.sha256`),
    bytes:Number(value.bytes??0),
  } as AssetDescriptor
}

export function parseManifest(value:unknown):StockScoutManifest{
  if(!isRecord(value))throw new Error('Manifest is not an object')
  if(value.manifestVersion===1){
    if(value.schemaVersion!==STOCKSCOUT_SCHEMA)throw new Error(`Unsupported schema ${String(value.schemaVersion??'unknown')}`)
    const assets=isRecord(value.assets)?value.assets:{}
    const counts=isRecord(value.counts)?value.counts:{}
    const status=value.status
    if(status!=='healthy'&&status!=='degraded'&&status!=='failed')throw new Error(`Unsupported scan status ${String(status)}`)
    return{
      ...value,
      manifestVersion:1,
      schemaVersion:STOCKSCOUT_SCHEMA,
      runId:requiredText(value.runId,'runId'),
      sessionDate:requiredText(value.sessionDate,'sessionDate'),
      marketDataDate:requiredText(value.marketDataDate??value.sessionDate,'marketDataDate'),
      generatedAt:requiredText(value.generatedAt,'generatedAt'),
      status,
      priceMode:requiredText(value.priceMode,'priceMode'),
      counts:{
        universe:Number(counts.universe??Number(counts.candidates??0)+Number(counts.excluded??0)),
        candidates:Number(counts.candidates??0),
        excluded:Number(counts.excluded??0),
        failed:Number(counts.failed??0),
        total:Number(counts.total??Number(counts.candidates??0)+Number(counts.excluded??0)),
      },
      provenance:isRecord(value.provenance)?value.provenance:{},
      versions:isRecord(value.versions)?value.versions as ScanManifestV1['versions']:{ranking:'unknown',detectors:'unknown',tradePlan:'unknown'},
      assets:{
        ...assets,
        core:requiredAsset(assets.core,'core'),
        details:requiredAsset(assets.details,'details'),
        excluded:requiredAsset(assets.excluded,'excluded'),
        history:requiredAsset(assets.history,'history'),
      },
    } as ScanManifestV1
  }
  if(value.manifestVersion===2){
    const assets=isRecord(value.assets)?value.assets:{}
    return{
      ...value,
      manifestVersion:2,
      model:requiredText(value.model,'model'),
      generatedAt:requiredText(value.generatedAt,'generatedAt'),
      universe:Number(value.universe??0),
      provenance:isRecord(value.provenance)?value.provenance:{},
      assets:{
        ...assets,
        core:requiredAsset(assets.core,'core'),
        legacyIndex:requiredAsset(assets.legacyIndex,'legacyIndex'),
        legacyDetails:requiredAsset(assets.legacyDetails,'legacyDetails'),
        legacyConfirmation:requiredAsset(assets.legacyConfirmation,'legacyConfirmation'),
        charts:requiredAsset(assets.charts,'charts'),
      },
    } as LegacyManifestV2
  }
  throw new Error(`Unsupported manifest v${String(value.manifestVersion??'unknown')}`)
}

export function normalizeCore(value:unknown,manifest:StockScoutManifest):CandidateCoreV1{
  if(!isRecord(value))throw new Error('Core dataset is not an object')
  const universe=Array.isArray(value.universe)?value.universe:Array.isArray(value.candidates)?value.candidates:null
  if(!universe)throw new Error('Core dataset has no candidate universe')
  const generatedAt=requiredText(value.generatedAt,'core.generatedAt')
  if(generatedAt!==manifest.generatedAt)throw new Error('Core dataset does not match manifest timestamp')
  const expected=manifest.manifestVersion===1?manifest.counts.candidates:manifest.universe
  if(universe.length!==expected)throw new Error(`Core candidate count ${universe.length} does not match manifest ${expected}`)
  const runId=manifest.manifestVersion===1?manifest.runId:manifest.runId
  return{
    ...value,
    generatedAt,
    runId:typeof value.runId==='string'?value.runId:runId,
    market:isRecord(value.market)?value.market:{},
    universe:universe.map((candidate,index)=>{
      if(!isRecord(candidate))throw new Error(`Candidate ${index} is invalid`)
      const ticker=requiredText(candidate.ticker,`candidate ${index}.ticker`).toUpperCase()
      return{
        ...candidate,
        id:typeof candidate.id==='string'?candidate.id:`scan:${runId??generatedAt}:candidate:${ticker}`,
        ticker,
        scanOrder:Number(candidate.scanOrder??index),
      } as CandidateSummaryV1
    }),
  } as CandidateCoreV1
}

export function isManifestV1(manifest:StockScoutManifest):manifest is ScanManifestV1{
  return manifest.manifestVersion===1
}

export function detailShardFor(ticker:string,count=128){
  const normalized=ticker.trim().toUpperCase()
  let value=0
  for(let index=0;index<normalized.length;index++)value+=(index+1)*normalized.charCodeAt(index)
  return String(value%Math.max(1,Math.floor(count))).padStart(3,'0')
}
